pkg/cmd/roachtest/datadog/datadog.go

```diff
@@ -0,0 +1,461 @@
+	}
+	return false
+}
+
+// MaybeUploadTestLog reads the log at logPath and transforms each
+// well-formatted line into Datadog HTTPLogItem JSON entries, then uploads them
+// to Datadog in batches. Uses a worker pool to parallelize network I/O and not
+// hold the entire log in memory.
+// Return an error if datadog credentials are not set.
+func MaybeUploadTestLog(
+	ctx context.Context, l *logger.Logger, logPath string, cfg LogMetadata,
+) error {
+
+	datadogLogger, err := l.ChildLogger("datadog", logger.QuietStdout)
+	if err != nil {
+		return err
+	}
+	datadogCtx, err := newDatadogContext(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Create Datadog Logs API client
+	// Default timeout is 60 seconds per HTTP request
+	configuration := datadog.NewConfiguration()
+	apiClient := datadog.NewAPIClient(configuration)
+	logsAPI := datadogV2.NewLogsApi(apiClient)
+
+	totalEntries, err := parseAndUploadLogFile(datadogCtx, datadogLogger, logsAPI, logPath, cfg)
+	if err != nil {
+		return errors.Wrapf(err, "failed to parse and upload log file: %s, uploaded %d log entries",
+			logPath, totalEntries)
+	}
+	l.Printf("successfully uploaded %d log entries to Datadog", totalEntries)
+	return nil
+}
+
+// newDatadogContext returns a derived datadog context from the provided context
+// The context requires DD_API_KEY and DD_SITE to be set.
+func newDatadogContext(ctx context.Context) (context.Context, error) {
+
+	// Use DD_SITE and DD_API_KEY env vars if they are set, otherwise look for
+	// values in roachtestflags and then set those values to the expected env
+	// vars.
+	if datadogSite := os.Getenv(envDatadogSite); datadogSite == "" {
+		if datadogSite = roachtestflags.DatadogSite; datadogSite == "" {
+			return nil, errors.Newf("%s env var or datadog-site flag is required", envDatadogSite)
+		}
+		if err := os.Setenv(envDatadogSite, datadogSite); err != nil {
+			return nil, err
+		}
+	}
+	if datadogAPIKey := os.Getenv(envDatadogAPIKey); datadogAPIKey == "" {
+		return nil, errors.Newf(
+			"%s env var is required", envDatadogAPIKey)
+	}
+	return datadog.NewDefaultContext(ctx), nil
+}
+
+// parseAndUploadLogFile reads and parses the log file serially. When a batch
+// is full, it is sent to a worker pool for upload to parallelize network I/O.
+// Best effort, a single worker uploading to datadog failing will not cancel
+// the entire upload.
+// Returns the total number of entries uploaded.
+func parseAndUploadLogFile(
+	ctx context.Context,
+	l *logger.Logger,
+	logsAPI *datadogV2.LogsApi,
+	logPath string,
+	logMeta LogMetadata,
+) (int, error) {
+
+	const numWorkers = 10
+	const batchSize = 1000 // Datadog max batch size
+
+	g := ctxgroup.WithContext(ctx)
+	batches := make(chan []datadogV2.HTTPLogItem, 20) // Buffered channel for batches
+	var totalUploaded atomic.Int64
+	var totalFailed atomic.Int64
+
+	// Start worker goroutines to upload batches concurrently
+	for i := 0; i < numWorkers; i++ {
+		g.GoCtx(func(ctx context.Context) error {
```

Comment:
nit: you could pass in a `t.NewGroup` at the original call site in `runTest`. Would make this resilient to unrecovered panics, although my initial reading of the code it doesn't seem like a big concern.